C wrapper generator for R's .Call interface

ptyszko opened the pull request:
It's likely that the .C interface will also need its own C wrappers. This PR does not consider them.

grzanka commented on scripts/ctype_translator.py:
```diff
@@ -14,8 +18,89 @@
+    c_void_p: 'raw'
+})
+
+
```
```suggestion
# comprehensive list of R types can be found here: https://cran.r-project.org/doc/manuals/r-release/R-ints.html
# in short:
# - NILSXP : NULL
# - CHARSXP : internal character strings
# - LGLSXP : logical vectors
# - INTSXP : integer vectors
# - REALSXP : numeric vectors
# - RAWSXP : raw vector
```

Provide a documentation in comment about the datatypes in R

ptyszko replied:
Added with extra explaination on NILSXP and RAWSXP

grzanka commented on scripts/ctype_translator.py:
```diff
@@ -14,8 +18,89 @@
+        return declarations_conversion[type_]
+    elif '*' in type_:
+        return POINTER(str_to_ctype(type_.replace('*', '', 1)))
+    elif ' int' in type_:
```
Add a comment explaining why checking for ` ' int'` and not for `'int'`

ptyszko replied:
Added comment with reasoning as well as more understandable way of checking

grzanka commented on scripts/ctype_translator.py:
```diff
@@ -14,8 +18,89 @@
+        return declarations_conversion[type_]  # will raise a KeyError that we'll want to catch later
+
+
 for ctype in list(mapping.keys()):
```
Isn't it enough like that ?
```suggestion
for ctype in mapping.keys():
```

ptyszko replied:
Copying the keys is necessary as we change the contents of `mapping` in the loop

grzanka commented on scripts/ctype_translator.py:
```diff
@@ -14,8 +18,89 @@
+        return declarations_conversion[type_]  # will raise a KeyError that we'll want to catch later
+
+
 for ctype in list(mapping.keys()):
```
Add a comment explaining why the loop below is needed

grzanka commented on scripts/header_parser.py:
```diff
@@ -433,13 +449,119 @@ def generate_dot_Call_wrapper(func_name: str, params_list: list[Parameter]) -> s
+    for param in params_list:
+        signature.append(f'\tSEXP p_{param.name},')
+        if param.size is None:
+            if (not param.raw_type.endswith('*')
```
I'd suggest to remove brackets inside if and merge the two lines below into single line

grzanka commented on scripts/header_parser.py:
```diff
@@ -452,19 +574,21 @@ def create_wrappers_for_header_file(path: str,
+        try:
+            absolute_path = Path(path).absolute()
+            name, params = parse_function_info(func, absolute_path)
+            if namespace is None or name in namespace:
```
# in case NAMESPACE file is not provided (here `not namespace`) we generate wrappers for all functions
# otherwise wrappers are generated only for functions listed in NAMESPACE file (here `name in namespace`)
```suggestion
            if not namespace or name in namespace:
```

ptyszko replied:
An empty namespace (e.g. if we create wrappers but don't want to export any of them) is also falsly, hence `namespace is None`, I will add the comment though

grzanka reviewed (changes requested):
please reply to the comments

ptyszko commented:
Added suggested comments and made suggested changes where applicable

grzanka commented:
@ptyszko what is the minimum required python version ?

```
(rege) (base) ➜  ReGenerator git:(wrapper-generator-C-wrapper) python scripts/header_parser.py --help
Traceback (most recent call last):
  File "scripts/header_parser.py", line 92, in <module>
    from typing import Union, Literal
ImportError: cannot import name 'Literal' from 'typing' (/home/grzanka/anaconda3/lib/python3.7/typing.py)
```

ptyszko commented:
3.8 for type hints. Could be lower if I removed them, but I would prefer not to

grzanka commented:
> 3.8 for type hints. Could be lower if I removed them, but I would prefer not to

can you add a readme information about that, otherwise code looks OK

grzanka reviewed (approved):
ok

ptyszko commented:
> > 3.8 for type hints. Could be lower if I removed them, but I would prefer not to
> 
> can you add a readme information about that, otherwise code looks OK

Added.
